torchao/prototype/smoothquant/api.py

```diff
@@ -0,0 +1,202 @@
+    recipe = torch.load(recipe_path, weights_only=True)
+
+    def recurse(module: torch.nn.Module, name: str = ''):
+        if isinstance(module, SmoothQuantObservedLinear):
+            smoothing_factor = recipe.get(name + ".smoothing_factor", None)
+            act_scales = recipe.get(name + ".act_scales", None)
+            wei_scales = recipe.get(name + ".wei_scales", None)
+            if device is not None:
+                module.to(device=device)
+            # act_scales is None for dynamic quantization
+            if any(x is None for x in (smoothing_factor, wei_scales)):
+                return module
+            return smooth_quant(smoothing_factor, act_scales, wei_scales)(module)
+
+        mod_new = module
+
+        for child_name, child in module.named_children():
+            full_name = f"{name}.{child_name}" if name else child_name
+            setattr(mod_new, child_name, recurse(child, full_name))
+        return mod_new
+
+    recurse(model)
+
+
+# StaticQuantizeAct and DynamicQuantizeAct are defined as classes to allow for easy serialization and deserialization
+class StaticQuantizeAct:
+    def __init__(self, factor, x_scale, target_dtype, quant_min=-127):
+        super().__init__()
+        self.factor = factor
+        self.x_scale = x_scale
+        self.x_zp = torch.zeros_like(x_scale, dtype=torch.int64)
+        self.target_dtype = target_dtype
+        self.quant_min = quant_min
+
+    def __call__(self, input):
+        x_zp = torch.zeros([1], dtype=torch.int64)
+        act = input / self.factor
+        act = act.to(input.dtype)
+        return to_affine_quantized_intx_static(act, self.x_scale, x_zp, list(act.shape), self.target_dtype, self.quant_min)
+
+
+class DynamicQuantizeAct:
+    def __init__(self, factor, target_dtype, quant_min=-127):
+        self.factor = factor
+        self.target_dtype = target_dtype
+        self.quant_min = quant_min
+
+    def __call__(self, input):
+        act = input / self.factor
+        act = act.to(input.dtype)
+        return to_affine_quantized_intx(act, MappingType.SYMMETRIC, _get_per_token_block_size(act), self.target_dtype, self.quant_min)
+
+
+def smooth_quant(
+        smoothing_factor: Optional[torch.Tensor] = None,
+        act_scales: Optional[torch.Tensor] = None,
+        wei_scales: Optional[torch.Tensor] = None
+    ):
+    """
+    Quantizes linear layers when passed into quantize_()
+
+    Args:
+        smoothing_factor: The smoothing factor for the layer. Acquired from the layer's observer if None.
+        act_scales: The activation scales for the layer. Acquired from the layer's observer if None.
+        wei_scales: The weight scales for the layer. Acquired from the layer's observer if None.
+    """
+
+    def quantize_weight(observed_linear):
+        target_dtype = torch.int8
+        quant_min = _DTYPE_TO_QVALUE_BOUNDS[target_dtype][0]
+        quant_max = _DTYPE_TO_QVALUE_BOUNDS[target_dtype][1]
```

Comment:
these can be None btw if they are the default values for target_dtype

Reply:
Thanks. Now it's hard-coded to -127/127 because there is a check of quant_min/_max in AQT here: https://github.com/pytorch/ao/blob/e7b33bc91c831d10249c1222c8b4b667f18f28b7/torchao/dtypes/affine_quantized_tensor.py#L1420

Reply:
It's removed. Thanks.